File: MATTHEW/leetcode/AIcoding/1.Gemini-2.5/rust_solutions/LCR 109-zlDJc7.rs
use std::collections::{VecDeque, HashSet};

impl Solution {
    pub fn open_lock(deadends: Vec<String>, target: String) -> i32 {
        let mut deadends_set: HashSet<String> = deadends.into_iter().collect();

        if deadends_set.contains("0000") {
            return -1;
        }
        if target == "0000" {
            return 0;
        }

        let mut queue: VecDeque<(String, i32)> = VecDeque::new();
        queue.push_back(("0000".to_string(), 0));

        let mut visited: HashSet<String> = HashSet::new();
        visited.insert("0000".to_string());

        while let Some((current_lock, steps)) = queue.pop_front() {
            for i in 0..4 {
                let mut chars: Vec<char> = current_lock.chars().collect();
                let digit = chars[i].to_digit(10).unwrap() as i32;

                // Increment digit
                let next_digit_inc = ((digit + 1) % 10).to_string().chars().next().unwrap();
                let mut next_chars_inc = chars.clone();
                next_chars_inc[i] = next_digit_inc;
                let next_lock_inc: String = next_chars_inc.into_iter().collect();

                if !visited.contains(&next_lock_inc) && !deadends_set.contains(&next_lock_inc) {
                    if next_lock_inc == target {
                        return steps + 1;
                    }
                    visited.insert(next_lock_inc.clone());
                    queue.push_back((next_lock_inc, steps + 1));
                }

                // Decrement digit
                let next_digit_dec = ((digit - 1 + 10) % 10).to_string().chars().next().unwrap();
                let mut next_chars_dec = chars; // Use the original chars again
                next_chars_dec[i] = next_digit_dec;
                let next_lock_dec: String = next_chars_dec.into_iter().collect();

                if !visited.contains(&next_lock_dec) && !deadends_set.contains(&next_lock_dec) {
                    if next_lock_dec == target {
                        return steps + 1;
                    }
                    visited.insert(next_lock_dec.clone());
                    queue.push_back((next_lock_dec, steps + 1));
                }
            }
        }

        -1
    }
}

File: leetcode/AIcoding/3.DeepSeek-v3/cpp_solutions/LCR 109-zlDJc7.cpp
#include <vector>
#include <string>
#include <queue>
#include <unordered_set>

using namespace std;

class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        unordered_set<string> dead(deadends.begin(), deadends.end());
        if (dead.find("0000") != dead.end()) {
            return -1;
        }
        if (target == "0000") {
            return 0;
        }

        unordered_set<string> visited;
        visited.insert("0000");
        queue<string> q;
        q.push("0000");
        int steps = 0;

        while (!q.empty()) {
            steps++;
            int size = q.size();
            for (int i = 0; i < size; i++) {
                string current = q.front();
                q.pop();
                for (int j = 0; j < 4; j++) {
                    for (int k = -1; k <= 1; k += 2) {
                        string next = current;
                        next[j] = (current[j] - '0' + k + 10) % 10 + '0';
                        if (next == target) {
                            return steps;
                        }
                        if (dead.find(next) == dead.end() && visited.find(next) == visited.end()) {
                            visited.insert(next);
                            q.push(next);
                        }
                    }
                }
            }
        }
        return -1;
    }
};